shell/browser/api/electron_api_safe_storage.cc

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2021 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/api/electron_api_safe_storage.h"
+
+#include "components/os_crypt/os_crypt.h"
+#include "shell/browser/browser.h"
+#include "shell/common/gin_converters/base_converter.h"
+#include "shell/common/gin_converters/callback_converter.h"
+#include "shell/common/gin_helper/dictionary.h"
+#include "shell/common/gin_helper/object_template_builder.h"
+#include "shell/common/node_includes.h"
+#include "shell/common/platform_util.h"
+
+namespace electron {
+
+namespace api {
+
+#if DCHECK_IS_ON()
+bool SafeStorage::electron_crypto_ready = false;
+#endif
+
+gin::WrapperInfo SafeStorage::kWrapperInfo = {gin::kEmbedderNativeGin};
+
+electron::api::SafeStorage::SafeStorage() {}
+electron::api::SafeStorage::~SafeStorage() = default;
+
+bool SafeStorage::IsEncryptionAvailable() {
+  return OSCrypt::IsEncryptionAvailable();
+}
+
+v8::Local<v8::Value> SafeStorage::EncryptString(v8::Isolate* isolate,
+                                                const std::string& plaintext) {
+  DCHECK(SafeStorage::electron_crypto_ready);
+  std::string ciphertext;
+  bool encrypted = OSCrypt::EncryptString(plaintext, &ciphertext);
+
+  if (!encrypted) {
+    gin_helper::ErrorThrower(isolate).ThrowError(
+        "Error while encrypting the text provided to "
+        "safeStorage.encryptString.");
+    return v8::Local<v8::Value>();
+  }
+
+  return node::Buffer::Copy(isolate, ciphertext.c_str(), ciphertext.size())
+      .ToLocalChecked();
+}
+
+std::string SafeStorage::DecryptString(v8::Isolate* isolate,
+                                       v8::Local<v8::Value> buffer) {
+  if (!node::Buffer::HasInstance(buffer)) {
```

Comment:
We should accept any `Uint8Array` here, and not require `Buffer` (which is a subclass of `Uint8Array`).

Reply:
I think `Buffer` is the better move here, as you note `Buffer` is a loose wrapper around `Uint8Array`, in fact you can do zero-cost conversions into a `Buffer` class or just pass a `Uint8Array` to anything that supports a `node::Buffer`.

Node documents things as taking Buffer's even if they also support `Uint8Array`'s.

Reply:
`node::Buffer::HasInstance` will return `false` for a `Uint8Array` though?

Reply:
@MarshallOfSound  @nornagon  Sorry to gravedig this one, just wondering if there was any resolution to this conversation/ and if there were any strong feelings against keeping `Buffer` as the input? 🙇

Reply:
I think for consistency we should use buffer, we don't have any other apis that take uint8array but we do work with buffers a bunch.

Reply:
I still think we should accept `Uint8Array` even if we document this as accepting `Buffer`. https://en.wikipedia.org/wiki/Robustness_principle